SCons/Tool/ninja.py

```diff
@@ -0,0 +1,1570 @@
+            result = None
+
+        NINJA_STAT_MEMO[path] = result
+        return result
+
+
+def ninja_noop(*_args, **_kwargs):
+    """
+    A general purpose no-op function.
+
+    There are many things that happen in SCons that we don't need and
+    also don't return anything. We use this to disable those functions
+    instead of creating multiple definitions of the same thing.
+    """
+    return None
+
+
+def ninja_whereis(thing, *_args, **_kwargs):
+    """Replace env.WhereIs with a much faster version"""
+    global NINJA_WHEREIS_MEMO
+
+    # Optimize for success, this gets called significantly more often
+    # when the value is already memoized than when it's not.
+    try:
+        return NINJA_WHEREIS_MEMO[thing]
+    except KeyError:
+        # We do not honor any env['ENV'] or env[*] variables in the
+        # generated ninja ile. Ninja passes your raw shell environment
+        # down to it's subprocess so the only sane option is to do the
+        # same during generation. At some point, if and when we try to
+        # upstream this, I'm sure a sticking point will be respecting
+        # env['ENV'] variables and such but it's actually quite
+        # complicated. I have a naive version but making it always work
+        # with shell quoting is nigh impossible. So I've decided to
+        # cross that bridge when it's absolutely required.
+        path = shutil.which(thing)
+        NINJA_WHEREIS_MEMO[thing] = path
+        return path
+
+
+def ninja_always_serial(self, num, taskmaster):
+    """Replacement for SCons.Job.Jobs constructor which always uses the Serial Job class."""
+    # We still set self.num_jobs to num even though it's a lie. The
+    # only consumer of this attribute is the Parallel Job class AND
+    # the Main.py function which instantiates a Jobs class. It checks
+    # if Jobs.num_jobs is equal to options.num_jobs, so if the user
+    # provides -j12 but we set self.num_jobs = 1 they get an incorrect
+    # warning about this version of Python not supporting parallel
+    # builds. So here we lie so the Main.py will not give a false
+    # warning to users.
+    self.num_jobs = num
+    self.job = SCons.Job.Serial(taskmaster)
+
+
+class NinjaNoResponseFiles(SCons.Platform.TempFileMunge):
+    """Overwrite the __call__ method of SCons' TempFileMunge to not delete."""
+
+    def __call__(self, target, source, env, for_signature):
+        return self.cmd
+
+    def _print_cmd_str(*_args, **_kwargs):
+        """Disable this method"""
+        pass
+
+
+def exists(env):
+    """Enable if called."""
+
+    # This variable disables the tool when storing the SCons command in the
+    # generated ninja file to ensure that the ninja tool is not loaded when
+    # SCons should do actual work as a subprocess of a ninja build. The ninja
+    # tool is very invasive into the internals of SCons and so should never be
+    # enabled when SCons needs to build a target.
+    if env.get("__NINJA_NO", "0") == "1":
```

Comment:
I wasn't sure exactly what the goal of this variable was, but seemed like it was to disable the ninja build if you invoked the build with ninja, which seemed an odd workflow?

Reply:
It's a way for ninja to run SCons and indicate to the ninja generator it should not disable building.

The way the Ninja generator works is it prevents SCons builders from actually executing (importantly this includes FunctionActions). For FunctionActions / Subst builders the generated ninja file will re-invoke SCons to build those targets. However because we carte blanche copy the SCons flags and vars the generator was invoked with this would mean the generator would still invade and disable this behaviour. So if you had a target `some-substed-file.txt` then ninja will invoke SCons to build that file however if the ninja generator is on `some-substed-file.txt` would never be built.

So we inject this in the generated SCons commands in the ninja file so the ninja generator knows that it's being invoked as a sub-process of ninja and doesn't disable building.